lms/services/canvas_api.py

```diff
@@ -91,10 +94,25 @@ def public_url(self, file_id):
 
         :rtype: str
         """
-        return self._helper.validated_response(
-            self._helper.public_url_request(self._oauth2_token.access_token, file_id),
+        oauth2_token = self._oauth2_token
+
+        return self.send_with_refresh_and_retry(
+            self._helper.public_url_request(oauth2_token.access_token, file_id),
             CanvasPublicURLResponseSchema,
-        ).parsed_params["public_url"]
+            oauth2_token.refresh_token,
+        )["public_url"]
+
+    def send_with_refresh_and_retry(self, request, schema, refresh_token):
+        try:
+            return self._helper.validated_response(request, schema).parsed_params
```

Comment:
Is the `expires_in` checked before the request is actually sent? If not, we might be able to save a few hundred ms by not even sending the initial request if we know the token has definitely expired.

Reply:
Nope, although we record `expires_in` (and also the `received_at` needed to interpret it) we never currently make use of it. I agree that this would be a simple and worthwhile optimisation to add because the common case of an expired access token (Canvas access tokens last for one hour) results in quite a few network hops:

1. The XHR request from the browser to our server and back
2. Our server's first attempt to send the Canvas API request and wait for the response
3. Refresh request and response
4. Second try at the original request and response

So the top-level XHR request here could be quite slow to respond, especially since some Canvas API's are pretty slow, and eliminating one of the network round-trips here should be worthwhile.